ick/runner.py

```diff
@@ -59,33 +62,71 @@ def iter_rule_impl(self) -> Iterable[BaseRule]:
             i = get_impl(rule)(rule, self.rtc)
             yield i
 
-    def test_rules(self) -> None:
-        with ThreadPoolExecutor() as tpe, Progress() as progress:
-            outstanding = {}
-            prepare_key = progress.add_task("Prepare", total=None)
+    def test_rules(self) -> int:
+        """
+        Returns an exit code (0 on success)
+        """
+        with ThreadPoolExecutor() as tpe:
+            print("[dim]testing...[/dim]")
+            # It's about this point I realize that yes, I am writing a whole
+            # test runner and that's not the business I want to be in.  Oh
+            # well...
+            buffered_output = []
```

Comment:
This should be a StringIO to start with, it will simplify the traceback handling.